### ISSUE TITLE
dev Environment broken when Using FastCGI 

### ISSUE THREAD

**Issue by DanielSchwiperich** [contao/core-bundle#426]
The frontend gets broken when using the dev environment `app_dev.php` for frontend pages on systems where php is running as fcgi.

You can take a look here where we working on the contao 4 port of our themo2go. Added a tmp access:
http://test:test@4.0-contao.demos.dev.cts-media.eu/app_dev.php/gallery.html

Throws a 404 error, while
http://4.0-contao.demos.dev.cts-media.eu/gallery.html
works fine. 

This is caused by [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117)

On fcgi servers the var `$_SERVER['PATH_INFO']` is used instead of `$_SERVER['SCRIPT_NAME']`
which is used on servers where apache for example uses php as module.

The problem that `script_name` is set wrong on cgi environments should be fixed since php 5.2:
`cgi.fix_pathinfo 	"1" 	PHP_INI_SYSTEM 	Verfügbar seit PHP 4.3.0. Vor PHP 5.2.1 PHP_INI_ALL.`
http://php.net/manual/de/ini.core.php#ini.cgi.fix-pathinfo

While PATH_INFO and SCRIPT_NAME contain  different values, they are equal on their system counterpart (cgi / apache mod)

*fcgi*
```
_SERVER["PATH_INFO"]	/_profiler/phpinfo
_SERVER["SCRIPT_NAME"]	/app_dev.php
```

*Apache 2.0 Handler* (apache mod) 
```
_SERVER["PATH_INFO"]	/_profiler/phpinfo
_SERVER["SCRIPT_NAME"]	/app_dev.php
```

The scriptName function ist also part of contao 3, so maybe an old leftover which is no longer needed.

**Comment by DanielSchwiperich** [comment]
Forgot to add: When we force using $_SERVER["SCRIPT_NAME"] on the fcgi systems it does work and we couldn't notice any other misbehavior

**Comment by leofeyer** [comment]
The two examples above are exactly the same, aren't they?

**Comment by DanielSchwiperich** [comment]
Yes, output is the same (and should be the same since php 5.2)

Just saw the first link had a typo:
http://test:test@4.0-contao.demos.dev.cts-media.eu/app_dev.php/gallery.html

We could reproduce this on different servers with different installations. 
The example uses Contao 4.0.4 but it's the same with 4.1.0 

I can give you access to that test environment if you like. It's just a simple Digital Ocean Test Env.
Should be the same for cgi and fpm, not sure about isapi though (never used that )

**Comment by aschempp** [comment]
I think @leofeyer did mean your *fcti* and *Apache2* examples are the same… ??

**Comment by DanielSchwiperich** [comment]
Hmm not sure I understand. 
There are not 2 example Setups I posted. Just one. 
http://test:test4.0-contao.demos.dev.cts-media.eu/app_dev.php/gallery.html
http://4.0-contao.demos.dev.cts-media.eu/gallery.html
It's the same fcgi environment. The bug only occurs when running dev mode because of the relative path `app_dev.php/` which triggers `Contao\Environment::scriptName()` to be called. 

I assumed most of you possibly do not run on apache + php fcgi (should be the same on cgi, fpm) > 5.3, otherwise you should already stumbled over it as well. 

When we found the bug we saw it's environment specific, meaning only in dev mode (app_dev.php) + fcgi (cgi / fpm as mentionen aswell) but not for setups where php is running as apache mod. 
It's pretty obvious when you take a look into [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117) 
I also showed with the php.net link why this had been correct in the past but not now anymore.

**Comment by leofeyer** [comment]
@aschempp No, I mean this:

> While PATH_INFO and SCRIPT_NAME contain  different values, they are equal on their system counterpart (cgi / apache mod)
>
> *fcgi*
> ```
> _SERVER["PATH_INFO"]	/_profiler/phpinfo
> _SERVER["SCRIPT_NAME"]	/app_dev.php
> ```
>
> *Apache 2.0 Handler* (apache mod) 
> ```
> _SERVER["PATH_INFO"]	/_profiler/phpinfo
> _SERVER["SCRIPT_NAME"]	/app_dev.php
> ```

The two code snippets are exactly the same, aren't they?

**Comment by aschempp** [comment]
@leofeyer yeah that's what I meant :D

**Comment by leofeyer** [comment]
:innocent:

**Comment by DanielSchwiperich** [comment]
And if you meant this part: 

>  *fcgi*
> ```
> _SERVER["PATH_INFO"]	/_profiler/phpinfo
> _SERVER["SCRIPT_NAME"]	/app_dev.php
> ```
>
> *Apache 2.0 Handler* (apache mod) 
> ```
> _SERVER["PATH_INFO"]	/_profiler/phpinfo
> _SERVER["SCRIPT_NAME"]	/app_dev.php
> ```

Yes it's the same for both. But it hasn't been before php 5.3
Before `script_name` had the fcgi script path value and the workaround was to use `path_info`
On  fcgi systems [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117)  returns the value of `_SERVER["PATH_INFO"]`  
On other setups (php running as apache mod) `_SERVER["SCRIPT_NAME"]` (which is the correct one)
`_SERVER["PATH_INFO"]`  cannot be used anymore because the value now differs to `script_name` 
leading to an error like in the exampe link. 
If we overwrite the method and forcing `_SERVER["SCRIPT_NAME"]` as return values on those fcgi systems it then works

**Comment by DanielSchwiperich** [comment]
Before `cgi.fix_pathinfo = 1 ` was introduced the value in `script_name` was something like `/var/www/share/vhost/php-cgi.conf` on cgi systems instead of the real script path. 
That what I meant with *fcgi script path value*

**Comment by leofeyer** [comment]
> But it hasn't been before php 5.3

Contao does not support PHP 5.2 or earlier. It is PHP 5.3 up to Contao 3.4, PHP 5.4 for Contao 3.5 and PHP 5.5 for Contao 4.

**Comment by DanielSchwiperich** [comment]
And that's while it's broken now on Systems with php > 5.3 because [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117) is a workaround which was needed before php 5.3

**Comment by DanielSchwiperich** [comment]
In the example setup it's php 5.4: http://4.0-contao.demos.dev.cts-media.eu/info.php

**Comment by DanielSchwiperich** [comment]
I'll try to explain it in another way

The current implementation of [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117) is a workaround for fcgi systems. 
Why? Because `_SERVER["SCRIPT_NAME"]` did not contain the value `/website/index.php`, it did contain the cgi script path for example `var/www/share/vhost/php-cgi.conf`

So [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117) is not returning `_SERVER["SCRIPT_NAME"]` on fcgi systems, it's returning `_SERVER["PATH_INFO"]` because it was containing the *actual/correct* `script_name` value on fcgi systems.

```
/**
	 * Return the relative path to the script (e.g. /website/index.php)
	 *
	 * @return string The relative path to the script
	 */
	protected static function scriptName()
	{
		return (static::$strSapi == 'cgi' || static::$strSapi == 'isapi' || static::$strSapi == 'cgi-fcgi' || static::$strSapi == 'fpm-fcgi') && (@$_SERVER['ORIG_PATH_INFO'] ?: $_SERVER['PATH_INFO']) ? (@$_SERVER['ORIG_PATH_INFO'] ?: $_SERVER['PATH_INFO']) : (@$_SERVER['ORIG_SCRIPT_NAME'] ?: $_SERVER['SCRIPT_NAME']);
	}
```

In PHP 5.2.1 there was a php ini setting introduced `cgi.fix_pathinfo` with default value 1 to fix that problem. 
The PATH_INFO Variable is since then usually not longer available. `Contao\Environment::scriptName()` did still work because it has a fallback, returning SCRIPT_NAME if there is not PATH_INFO even on fcgi Systems. 

Take a look here at this fcgi setup: 
http://4.0-contao.demos.dev.cts-media.eu/info.php
No PATH_INFO set

NOW symfony, running in dev mode, does SET this var PATH_INFO (maybe it's needed for the debug toolbar).
This is causing `Contao\Environment::scriptName()` to now return the PATH_INFO variable instead of the SCRIPT_NAME
And because this two VARS do differ the dev mode gets unusable leading to 404 errors while the prod mode is running fine:
Broken dev mode: (link formatting broken by github/markdown)
http://test:test@4.0-contao.demos.dev.cts-media.eu/app_dev.php/gallery.html
Not broken prod mode:
http://4.0-contao.demos.dev.cts-media.eu/gallery.html

If you call the link: http://test:test@4.0-contao.demos.dev.cts-media.eu/app_dev.php/gallery.html
und click the php button in the toolbar you can see that PATH_INFO now exists. 

`_SERVER["PATH_INFO"]	/_profiler/phpinfo`
But containing a bad value leading to the 404 error in dev mode, 
because the value should be the same line SCRIPT_NAME
`_SERVER["SCRIPT_NAME"]	/app_dev.php` 

I added a phpinfo call at the end of fe_page. 
http://4.0-contao.demos.dev.cts-media.eu/gallery.html
You can see in prod mode scrolling down, PATH_INFO is not set, that's why prod mode is working.

**Conclusion**
Symfonies dev mode / debug toolbar does set a new Server Var `$_SERVER["PATH_INFO"]` leading `Contao\Environment::scriptName()` on fcgi systems to return the wrong value. 

The fix needs to be done in `Contao\Environment::scriptName()` imho because the workaround behavior should no longer be needed.

**Comment by leofeyer** [comment]
I don't think this is a bug. We are using the same code as e.g. TYPO3 does and this code has not changed in years.

https://forge.typo3.org/attachments/11831/class.t3lib_div.php#L3455

@contao/developers /cc

**Comment by DanielSchwiperich** [comment]
I did some more testing and pretty sure it's a bug and also not symfony (debug bar) related.

The PATH_INFO Server variable gets set only when needed by PHP or Apache.
If you call this link:
http://4.0-contao.demos.dev.cts-media.eu/test/info.php
no PATH_INFO
If you call this link:
http://4.0-contao.demos.dev.cts-media.eu/test/info.php/test123
PATH_INFO is set. 

 [`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117) 
tries to get the name of the running php script

If you've URL/webseite/index.php
It should be `/webseite/index.php`
If you something like URL/test/info.php/test123
It should be `/test/info.php`

Usually the function only returns the Server var SCRIPT_NAME. This is correct and works fine. 
There is one exception to it. On `'cgi' ||  'isapi' || 'cgi-fcgi' ||  'fpm-fcgi'` Systems the Server var PATH_INFO is returned if it's set.

PATH_INFO does not get set for something like URL/webseite/index.php because there is no PATH_INFO. 
See here: http://4.0-contao.demos.dev.cts-media.eu/test/info.php
No PATH_INFO

But if you call this link: http://4.0-contao.demos.dev.cts-media.eu/test/info.php/test123
it does get set and the value is NOT /test/info.php. It's test123.
And the value is correct by definition: http://php.net/manual/de/reserved.variables.server.php
But *NOT* appropriate to use in this function where the value should /test/info.php

----
PATH_INFO cannot be used as subsitute for SCRIPT_NAME and it's also not needed.

And yes the bug exists also in previous contao versions. I just did test this. 
It has not arised because nowadays nowone is running without mod rewrite I would think.

Here a contoa 3.2 system on fcgi:
http://www.cts-media.eu/warum-wir.html
works, but 
http://www.cts-media.eu/index.php/warum-wir.html
=> 404

Another contao website where php is running as module and subsequently always SCRIPT_NAME is used:
http://www.hd-konz.de/sportster-33.html
works, and also
http://www.hd-konz.de/index.php/sportster-33.html
works

----

I've this verified by 2 of our employees. If you don't believe me, then please set up a fcgi environment and test it by yourself. 
We did verify this on 3 completely different setups to make sure it's not just a server misbehavior (Custom Docker, Maxcluster, Digital Ocean)

----

It's not only a bug, but now with contao 4 also a problem because if we want to run the debug mode URLS look like URL/app_dev.php/seite.html leading to 404 errors where is running in non debug mode everything is fine URL/seite.html  (because we've no PATH_INFO var introduced)

----
tl;dr
Sorry for this text of wall again. I always try to be short and simple but it's not so easy in this case. 

To put in one sentence: 
[`Contao\Environment::scriptName()`](https://github.com/contao/core-bundle/blob/master/src/Resources/contao/library/Contao/Environment.php#L117) has a wrong implementation. In some cases it returns PATH_INFO instead of SCRIPT_NAME
To return the value of PATH_INFO is wrong by definition if you want the php script name.
Compare definitions here: http://php.net/manual/de/reserved.variables.server.php

Some years ago this was ok to use on setups running php lower than 5.2.1, but not on current systems. See first post.

**Comment by DanielSchwiperich** [comment]
Additional why this hasn't been a problem previous contao systems is because the implementation of the function Frontend::getPageIdFromUrl() has changed in Contao 4

In **Contao 4** right at the beginning of the function we've `$strRequest = \Environment::get('relativeRequest');` which does access the scriptName function 
[Contao 4 Frontend::getPageIdFromUrl()](https://github.com/contao/core-bundle/blob/b59a083a05b01ff95b7246ce6a734bf62a67919e/src/Resources/contao/classes/Frontend.php#L64) 

In **contao 3** the function does not get called at all, that's why stuff like URL/index.php/seite.html does work.
[Contao 3 Frontend::getPageIdFromUrl()](https://github.com/contao/core/blob/fba9fa1c29e4e5db418f9827bde20f2e4608ab5e/system/modules/core/classes/Frontend.php#L58)
There `Environment::get('request')` does get called which subsequently call `Enviroment::requestUri()`
And there scriptName would only be called if REQUEST_URI is not set, but this is a var which is always set in those cases.

**Comment by discordier** [comment]
I have to admit, this makes sense. Maybe it would be better to use the request stack to get the current request and fall back to previous behaviour when no request is avaliable.
No matter how we change it, the work around should not be needed at all anymore and therefore can be removed.
I will have to test it on another environment though.

**Comment by fritzmg** [comment]
I actually have the same problem in a pure Symfony setup. When running PHP under CGI on my local XAMPP, the routing fails completely when using `app_dev.php/…` (except for the default URL without parameters).

**Comment by leofeyer** [comment]
That's interesting. Might not be a Contao bug then.

**Comment by DanielSchwiperich** [comment]
Please see my last two posts here in this issue. They describe everything needed to solve this bug and also show it's a Contao Bug and not a Symfony bug.

If it would be a Symfony bug then it would have been fixed long time ago since fcgi, fpm environments are not unusual. 
I wonder when @discordier  finds time to validate it. The fix is only to change one line of code.

And just to be sure I just set up a docker container with fcgi, installed the symfony demo and it worked fine using `app_dev.php` + routes like `http://172.17.0.2/symfony_demo/web/app_dev.php/en/blog/posts/eros-diam-egestas-libero-eu-vulputate-risus` . I think there must be something else going at your setup @fritzmg

**Comment by fritzmg** [comment]
> I think there must be something else going at your setup @fritzmg

Yeah I think so too.

**Comment by leofeyer** [comment]
There is a [test for the Environment class](https://github.com/contao/core-bundle/blob/master/tests/Contao/EnvironmentTest.php#L83-L108), which sets up the `$_SERVER` array according to the different environments. Can you please dump `$_SERVER` on your system an post it here?

> NOW symfony, running in dev mode, does SET this var PATH_INFO (maybe it's needed for the debug toolbar).

Where did you find this?

**Comment by DanielSchwiperich** [comment]
hi leo, I've update the relevant post with functional not outdated links: 
https://github.com/contao/core-bundle/issues/426#issuecomment-167056629
This post also contains the latest infos and correct findings on this topic and has the  `$_SERVER` values you ask for. 

For the second part, this is also explained in the mentioned post. It not gets set by symfony, but by apache / php itself as soon as there is php file called with a path appended.

**Comment by leofeyer** [comment]
That's quite a weird setup. There a several server variables missing:

* `DOCUMENT_ROOT`
* `ORIG_SCRIPT_FILENAME`
* `ORIG_SCRIPT_NAME`
* `ORIG_PATH_INFO`
* `ORIG_PATH_TRANSLATED`

And the `REQUEST_URI` variable is incorrect, as it is the same as `SCRIPT_NAME` but should contain the real request string instead.

Why is this?

**Comment by leofeyer** [comment]
How does your FastCGI setup file look like (should be in `/etc/php5/cgi`)?

**Comment by DanielSchwiperich** [comment]
? 
`DOCUMENT_ROOT` is there, 
the mentioned ORIG_ vars do not exist by default according to http://php.net/manual/en/reserved.variables.server.php 
Nevertheless they only get set if not empty / differ from the current value afaik. 
`PATH_INFO` is explained in the mentioned post

`REQUEST_URI` also looks fine to me? Or what is wrong with:
`http://4.1-contao.demos.dev.cts-media.eu/test/info.php/test123?test=blabla` and value of `/test/info.php/test123?test=blabla` for it. 

`SCRIPT_NAME` is what the contao bug is about, since the workaround is not needed anymore.
See the OP.
> The problem that script_name is set wrong on cgi environments should be fixed since php 5.2:
cgi.fix_pathinfo "1" PHP_INI_SYSTEM Verfügbar seit PHP 4.3.0. Vor PHP 5.2.1 PHP_INI_ALL.
http://php.net/manual/de/ini.core.php#ini.cgi.fix-pathinfo

The Server should be a standard fcgi Setup done by maxcluster. Like mentioned we did confirm this on a standard digital ocean instance and a local docker instace too. 
I could setup a DO fcgi default environment for comparison again if you like. 

Here's the cgi ini you asked for. Only the basic ini without it's subparts.

fcgi wrapper: 
```
#!/bin/sh

export PHPRC="/etc/php5/cgi"
exec /usr/bin/php5-cgi -d memory_limit=2048M -d max_execution_time=1800
```

Subparts (not attached, just the listing)
`ll /etc/php5/cgi/conf.d/`
```
10-pdo.ini  20-apc.ini	20-curl.ini  20-gd.ini	20-intl.ini  20-mcrypt.ini  20-mysql.ini  20-mysqli.ini  20-pdo_mysql.ini
```
And the ini from /etc/php5/cgi/: 
[php.ini.zip](https://github.com/contao/core-bundle/files/231524/php.ini.zip)
Shoud be pretty much default if I'm not wrong.


hmm, did you delete ony of your post. I could swear there was another one

**Comment by leofeyer** [comment]
How does your Apache vhost and/or .htaccess file look like?

**Comment by DanielSchwiperich** [comment]
```
<VirtualHost *:80>

  ServerName 4.1-contao.demos.dev.cts-media.eu
  DocumentRoot /var/www/share/contao-projekte/demos/contao-4.X/htdocs/web
  SuexecUserGroup web-user web-user
  IPCCommTimeout 600
  Options None


  <Directory /var/www/share/contao-projekte/demos/contao-4.X/htdocs/web>
    FCGIWrapper /var/www/share/dev.cts-media.eu/php-cgi.conf .php
    <FilesMatch \.php$>
        SetHandler fcgid-script
    </FilesMatch>
    Options +ExecCGI +SymLinksIfOwnerMatch
    AllowOverride All
    Order allow,deny
    Allow from all
  </Directory>

  CustomLog ${APACHE_LOG_DIR}/4.1-contao.demos.dev.cts-media.eu/access.log combined
  ErrorLog  ${APACHE_LOG_DIR}/4.1-contao.demos.dev.cts-media.eu/error.log
  LogLevel warn

</VirtualHost>
```
Nothing changed in .htaccess (contao 4.1)

**Comment by leofeyer** [comment]
And `/var/www/share/dev.cts-media.eu/php-cgi.conf`?

**Comment by DanielSchwiperich** [comment]
See post above, for the wrapper. 
here again:
`/var/www/share/dev.cts-media.eu/php-cgi.conf`

```
#!/bin/sh

export PHPRC="/etc/php5/cgi"
exec /usr/bin/php5-cgi -d memory_limit=2048M -d max_execution_time=1800
```

**Comment by leofeyer** [comment]
Ah, ok. I wasn't sure about that.

**Comment by DanielSchwiperich** [comment]
Did take a quick look, in the `testCgiFcgi` Unittest. It's missing the case we got here. 

It's testing for  `core/index.php` which works fine but not for the case we got here for example `/app_dev.php/gallery.html` which is causing the problem when calling a php file with path appended. 

We would need another test for those case where Contao\Environment::scriptName() is triggered.

Also not sure about some of the server values in that test.

**Comment by fritzmg** [comment]
I also tested this on a server running Plesk with a Vhost configured for PHP `5.6.20` under FastCGI. When requesting `http://www.example.org/app_dev.php/test.hml`, the result of
```php
\Environment::get('relativeRequest');
```
is
```
app_dev.php/test.html
```
thus causing `Frontend::getPageIdFromUrl` to return `app_dev.php` instead of `test`.

**Comment by leofeyer** [comment]
@DanielSchwiperich Does the following fix the issue?

```php
protected static function scriptName()
{
	$request = \System::getContainer()->get('request_stack')->getCurrentRequest();

	if ($request === null)
	{
		return @$_SERVER['ORIG_SCRIPT_NAME'] ?: $_SERVER['SCRIPT_NAME'];
	}

	return $request->getScriptName();
}
```

**Comment by DanielSchwiperich** [comment]
Can not test atm, will do later or on the weekend. But using the symfony implementation should fix it, yes. Will report back

**Comment by fritzmg** [comment]
@leofeyer works in my test case :+1:

**Comment by fritzmg** [comment]
There is also a negative effect when running PHP under CGI (not FastCGI). When running under CGI (at least in my local Windows XAMPP setup), the `app_dev.php` generally works with URLs like
```
http://www.example.org/app_dev.php/test.html
```
However, if there is a navigation module present, the page "test" will not be marked as active when you are on it. 

When accessing the same page without `app_dev.php` or with `app_dev.php`, but with `mod_php`, it works. This was tested with Contao `4.1.3`.

After applying [@leofeyer's fix](https://github.com/contao/core-bundle/issues/426#issuecomment-213388407), the navigation module also works as expected with the `app_dev.php` under the CGI environment.

**Comment by DanielSchwiperich** [comment]
fix confirmed :smile: 
tested on a fresh 4.1.3 install.

**Comment by leofeyer** [comment]
Fixed in 22f0bde87a14103f2bcf6d87a2c3df293f77b41c.
